### ISSUE TITLE
images returned by S3Man are corrupted..

### ISSUE THREAD

**Issue by donileo** [aws/aws-sdk-ios#180]
For some reason when downloading images from AWS S3 via S3TransferManager I am getting images that look corrupted. Truly I have no idea why but I can confirm these are the images the SDK is returning. I am not modifying the images. For example here is the image as its stored in S3.. and below what gets returned "sometimes" when I ask S3TransferManager to download them. Sometimes it returns them as the original looks but plenty of other times it returns the corrupt looking copy you may see below or similar.

![place1_photo1](https://cloud.githubusercontent.com/assets/1881038/7218437/1b7ad5f8-e63d-11e4-8a60-c41f14e98ded.jpg)
![s3image](https://cloud.githubusercontent.com/assets/1881038/7218438/1b7fb74e-e63d-11e4-9d1a-8cc14cb22f94.jpg)

Update: It seems this is caused when S3TransferManager is asked to download the same file more than once. What it will do is a bit weird: 

 When file1 starts downloading it'll start writing to file1Path, no problem yet.. When file2 is asked to download it'll delete file1's currently written data (if file1 started downloading) and then proceed to start writing to file1Path. In this case since two transfers are occurring for the same file both transfer are writing simultaneously to the same file path leading to those image artifacts that you see.

**Comment by yyzhu** [comment]
Hi donileo,

Thank you for the report,I can confirm that it is  a bug in S3TransferManger and we will fix it in future release. 

In the mean time, there is a workaround: If you don't specify the downloadingFileURL in `AWSS3TransferManagerDownloadRequest.downloadingFileURL`,  S3TransferManager will automatically create a temporary filePath which can be retrieved later in `AWSS3TransferManagerDownloadOutput.body`, and then can be moved  to the original Filepath you wish to set  in downloadRequest.

For example:
```
AWSS3TransferManagerDownloadRequest *downloadRequest  = [AWSS3TransferManagerDownloadRequest new];
 downloadRequest.bucket = @"bucketname";
downloadRequest.key = @"keyname";

//comment out. Instead, using NSURL retrieved from  AWSS3TransferManagerDownloadOutput
//downloadRequest.downloadingFileURL = originalFileURL;

[[transferManager download:downloadRequest] continueWithBlock:^id(BFTask *task) {
            AWSS3TransferManagerDownloadOutput *output = task.result;
            NSURL *receivedBodyURL = output.body;
            
            [[NSFileManager defaultManager] moveItemAtURL:receivedBodyURL
                                                            toURL:originalFileURL
                                                            error:nil];

            return nil;
        }];

```

**Comment by donileo** [comment]
Hi, yyzhu, in regards to your proposed alternative workaround above doing what you are proposing above still corrupts the images. Basically S3TransferManager will start writing to a file at the specified path (in the solution above it will use the temporary file path generated) then when the next file asks to download it will for some odd reason generate the same temporary file path.. then when the first download is done it will move the item it is writing to the specified originalFileURL and the download for the second file will be corrupted as you just moved the file it was writing to.

Simply put, the above solution will still corrupt the file.

**Comment by yyzhu** [comment]
Hi @donileo , I did some investigation and found out  this issue only happened while multiple downloading tasks are running simultaneously assigned by a same downloadingFileURL. The S3 transferManager failed to detect such behavior thus result a corrupted file.

We will add a mechanism to detect such behavior therefore downloading tasks will get an error if assigned downloadingFileURL currently being used by other download tasks.   Meanwhile, It is always a good idea to cancel the previous tasks before start the exactly same task again. you can cancel any s3 transfer Manager task by calling `[downloadRequest cancel]`

**Comment by donileo** [comment]
Correct thats exactly what the problem is. Hopefully you get it fixed asap.

**Comment by yosuke-matsuda** [comment]
The AWS Mobile SDK for iOS 2.1.2 should address this issue. Please let us know if you keep encountering the issue. Thanks.

**Comment by annathomasQB** [comment]
Hi,

I am facing the same issue even with the latest SDK in place. I believe the issue happens mainly when in low network.

![image](https://cloud.githubusercontent.com/assets/2297073/14631033/f571da56-062e-11e6-8c99-e1bb2fa8c177.png)

**Comment by pradipvaghasiya** [comment]
This happens to me as well with low network. I am using version 2.4.2 
Please check

**Comment by karthiksaligrama** [comment]
tracking in #415

**Comment by annathomasQB** [comment]
Thanks for the update 👍
